### DISCUSSION TITLE
[토론] IoC는 객체지향적인 기법인가?

### DISCUSSION THREAD

**Discussion by sihyung92** [woowacourse-tobi-spring-study/tobi-vol1#24] [Show and tell]
발단 : https://github.com/woowacourse-tobi-spring-study/tobi-vol1/pull/3#issuecomment-831610998

### 제어의 역전에 대한 코다의 코멘트
> 우리의 어플리케이션 코드가 라이브러리처럼 되어버린 건가요!! 뭔가 느낌이 객체지향이랑은 좀 먼거 같기도 하고 ㅠㅠ 웨지는 어떻게 생각하나요?

### 웨지 의견
IoC가 객체지향적이냐? 는 말에 대해선 IoC에서 등장하는 '주도권 없음' 이란 말과 객체지향에서 말하는 '자율성' 사이의 용어혼동 때문이라고 생각해요. 하지만 IoC에서 이야기하는 '주도권 없음'은 다시 표현하면 생성 책임을 다른 객체로 분리했다, 라는 말이 되므로 그렇게 생각하면 충분히 객체지향적이라고 생각합니다 ㅎ_ㅎ

생성 책임을 해당 객체로부터 분리하는 게 객체지향적이냐? 고 논의가 넘어가면...... 재밌는 토론이 되겠네용 ㅎ_ㅎ

### 손너잘 코멘트에서 긁어온 관련 내용
> 다만 읽으면서 의심이 드는 부분이 있다면, IoC가 적용되는 순간 객체지향의 특성과 벗어나는 코드가 작성된다고 생각한다. 능동적인 객체간의 협력이 아닌, 수동적인 객체간의 강제로 만들어진 협력체... 객체지향이 아니라 객체지향을 흉내낸 느낌이랄까? 생각보다 마음에 들지는 않는다.

**Comment by Joyykim** [comment]
좋은 화두네요ㅎㅎ 일단 제 의견은 "IoC는 객체지향적이다"예요.
일단 IoC에서 말하는 능동, 수동은 객체지향에서 말하는 "책임을 가진 능동적인 객체"의 능동과 다르다고 생각해요.

### 객체지향에서의 능동/수동
능동적인 객체는 스스로 그 책임을 어떻게 수행할 지를 결정하죠.
수동적인 객체는 남이 어떻게 할지 일일이 지시하며 자료구조처럼 취급돼요.

### IoC에서의 능동/수동
어플리케이션 흐름의 제어권을 어플리케이션 코드가 가진다면 능동.
프레임워크가 가진다면 수동.

**흐름의 제어권**과 **책임의 수행**은 다르다고 생각합니다.
저는 오히려 객체들이 충분히 객체지향적이어서 자율적으로 책임을 수행하고, 높은 응집도/낮은 결합도를 가져야만 IoC가 가능하다고 생각해요!!

다른 분들 의견은 어떠신가요?

**Comment by xlffm3** [comment]
객체지향 프로그램(Object-Oriented Program)이란 무엇일까요? 간결하면서도 명확하게 요약하고 싶은데, 아직까지는 어렵네요. 우테코 레벨 1 과정 때 객체지향의 사실과 오해 등의 서적을 탐닉하면서 느낀 바로는, 객체지향적인 프로그램이란 역할과 책임을 가진 자율적인 객체들의 협력을 통해 공동의 목표(개발자가 의도한 시스템의 정상적인 구동)를 이룩하는 것이라고 생각해요.

그렇다면 여기서 말하는 자율성이란 개념은 무엇일까요? 자율성에 대한 해석에 따라 IoC가 OOP에 부합하는 패턴인지에 대한 의견이 갈릴 것이라고 생각해요. 현실 세계에서 자율이라 함은 외부의 지배나 구속을 받지 아니하고 스스로의 원칙에 따라 어떤 일을 하는 것을 의미하겠죠? 이러한 관점에서 바라볼 때, 자율적인 객체란

* 자신에게 주어진 책임과 역할을 수행하는데 필요한 의존 관계를 주입이 아닌 주도적으로 설정한다.
* 원하는 시점과 위치에서 인스턴스화될 수 있다.
  * 즉, 라이프사이클을 본인이 관장한다.

정도의 특징을 가질 수 있을 거라고 생각해요. 우리가 학습한 IoC 컨테이너를 기반으로 DI를 통해 런타임 시점에 의존성이 결정되는 것이 아닌, 컴파일타임 시점의 코드에서 의존성이 드러나는 것이 오히려 객체가 자율적인 것이 아닌가 하는 기분이 들기도 합니다.

하지만 자율적인 객체라는 개념은 '성공적인 어플리케이션 구동'이라는 거시적인 차원의 목표 하에 비로소 의미를 가진다고 생각해요. 성공이라는 개념 역시 다양하게 해석될 수 있겠지만, 저는 미래의 변화에 유연하게 대처할 수 있음 정도로 번역하고 싶네요. 객체가 스스로 의존성을 설정하지 못하고 주입받는다면 그게 자율적이고 능동적인 것이 아니라, 의존적이고 수동적인 것일까요? (작성중입니다...)

**Reply by Joyykim** [reply]
5일째 작성중...😳
